include/merlin/core_kernels/find_or_insert.cuh

```diff
@@ -21,6 +21,1253 @@
+          __pipeline_wait_prior(3);
+          CopyValue::lds_stg(rank, dst, src, dim);
+        } else {
+          VecV** sm_bucket_values_ptr = SMM::bucket_values_ptr(smem);
+          auto bucket_values_ptr =
+              sm_bucket_values_ptr[groupID * GROUP_SIZE + i - 2];
+          auto key_pos_cur = g.shfl(key_pos, i - 2);
+          VecV* dst = bucket_values_ptr + key_pos_cur * dim;
+          __pipeline_wait_prior(3);
+          CopyValue::lds_stg(rank, dst, src, dim);
+        }
+        if (rank == i - 2) {
+          auto key_address = BUCKET::keys(bucket_keys_ptr, key_pos);
+          key_address->store(key, cuda::std::memory_order_release);
+        }
+      }
+    }
+  }
+  auto occupy_result_cur = g.shfl(occupy_result, GROUP_SIZE - 1);
+  uint32_t tx_cur = groupID * GROUP_SIZE + GROUP_SIZE - 1;
+  S* sm_param_scores = SMM::param_scores(smem);
+  S score_cur = scores == nullptr ? device_nano<S>() : sm_param_scores[tx_cur];
+  int** sm_buckets_size_ptr = SMM::buckets_size_ptr(smem);
+  auto bucket_size_ptr = sm_buckets_size_ptr[tx_cur];
+  __pipeline_wait_prior(1);
+  S* src = SMM::bucket_scores(smem, groupID, diff_buf(GROUP_SIZE));
+  while (occupy_result_cur == OccupyResult::INITIAL) {
+    int min_pos_local = -1;
+    S min_score_local = MAX_SCORE;
+#pragma unroll
+    for (int j = 0; j < BUCKET_SIZE; j += GROUP_SIZE * Load_LEN_S) {
+      S temp_scores[Load_LEN_S];
+      *reinterpret_cast<byte16*>(temp_scores) =
+          *reinterpret_cast<byte16*>(src + rank * Load_LEN_S + j);
+#pragma unroll
+      for (int k = 0; k < Load_LEN_S; k++) {
+        S temp_score = temp_scores[k];
+        if (temp_score < min_score_local) {
+          min_score_local = temp_score;
+          min_pos_local = rank * Load_LEN_S + j + k;
+        }
+      }
+    }
+    const S min_score_global = cg::reduce(g, min_score_local, cg::less<S>());
+    if (score_cur < min_score_global) {
+      if (rank == GROUP_SIZE - 1) {
+        occupy_result = OccupyResult::REFUSED;
+      }
+      occupy_result_cur = g.shfl(occupy_result, GROUP_SIZE - 1);
+      break;
+    }
+    uint32_t vote = g.ballot(min_score_local <= min_score_global);
+    if (vote) {
+      int src_lane = __ffs(vote) - 1;
+      int min_pos_global = g.shfl(min_pos_local, src_lane);
+      if (rank == GROUP_SIZE - 1) {
+        src[min_pos_global] = MAX_SCORE;  // Mark visited.
+        auto min_score_key = BUCKET::keys(bucket_keys_ptr, min_pos_global);
+        auto expected_key =
+            min_score_key->load(cuda::std::memory_order_relaxed);
+        if (expected_key != static_cast<K>(LOCKED_KEY) &&
+            expected_key != static_cast<K>(EMPTY_KEY)) {
+          auto min_score_ptr =
+              BUCKET::scores(bucket_keys_ptr, BUCKET_SIZE, min_pos_global);
+          bool result = min_score_key->compare_exchange_strong(
+              expected_key, static_cast<K>(LOCKED_KEY),
+              cuda::std::memory_order_acquire, cuda::std::memory_order_relaxed);
```

Comment:
Theoretically, `memory_order_acq_rel` should be more property for the success branch.

Reply:
I think memory_order_acquire is enough.
As we can gurantee that the modificatio to bucket can't be reorderd before the compare_exchange_strong()